### DISCUSSION TITLE
./gradlew :dependency --update-locks outputs too much text and cause "stdout maxBuffer exceeded" error

### DISCUSSION THREAD

**Discussion by ainame** [renovatebot/renovate#20736] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

I am helping with an Android app project to make Mend Renovate GitHub app work for the project. (I tried Self-hosted one for debugging too). The project has roughly 53 modules and `gradle.lockfile`. The app adopts modular architecture. When I run Renovate for the project, `./gradlew :dependency --write-locks` command with many modules specified in command line options fails with "stdout maxBuffer exceeded" error. This error was once reported and fixed in the past by increasing stdout maxBuffer to 10MB by default. 

https://github.com/renovatebot/renovate/issues/4971

However, when I measured the output size with our project using the same command as what Self-hosted Renovate executes locally, the output size became 59 MB. 

```
% ./gradlew --console=plain -q :dependencies :analytics:dependencies ... --update-locks androidx.preference:preference > gradle.log
% ls -lah gradle.log
-rw-r--r--  1 satoshinamai  staff     59MB  3 Mar 16:09 gradle.log
```

I don't think asking Renovate team to increase the default buffer size again is the way to go here. 

My proposal here is to do either of the following. I'm in favour of 1.

1. Completely drop stdout for  `./gradlew :dependencies --write-locks` by `> /dev/null`
2. Expose `maxBuffer` option to configurations

I think `stdout` for the command isn't useful for most cases, and Renovate users can't see the output anyway even if `LOG_LEVEL=debug`. (`stderr` should still be available for error reporting)

----

### Context/Background

After some research, it appears that the command `./gradlew :dependencies --write-locks` that Renvoate use is meant to visualise/render the dependency graph. This is why stdout maxBuffer can always be problematic for Gradle when a project is at scale.

> Gradle provides the built-in dependencies task to render a dependency tree from the command line.   

https://github.com/gradle/gradle/blob/master/subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/viewing_debugging_dependencies.adoc

This command was recommended in the official document. 

> When locking multiple configurations, you may want to lock them all at once, during a single build execution.
>  For this, you have two options:
> * Run gradle dependencies --write-locks. This will effectively lock all resolvable configurations that have locking enabled. Note that in a multi project setup, dependencies only is executed on one project, the root one in this case.
> * Declare a custom task that resolves all configurations. This does not work for Android projects.

https://docs.gradle.org/current/userguide/dependency_locking.html#generating_and_updating_dependency_locks

This is because `:dependencies` command and `--write-locks` option are isolated concepts. Gradle can update `lockfiles` with `--write-locks` option without `:depenencies` task, apparently. In this sense, you can define a custom task with the latter option in the above text. I think Renovate should still use the combination of `:dependencies` and `-write-locks` but drop `stdout` since Renovate would want to support only the standard way.

Gradle maintainer pointed out that the issue of too much output to stdout for `--write-locks` is caused by `:dependencies` in this ticket. 

> The output problem is not related to using --write-locks but to using dependencies as the trigger.

https://github.com/gradle/gradle/issues/21445

**Comment by ainame** [comment]
I'm not sure if there's a downside in dropping stdout for updating lockfile commands, so I might be wrong😓

**Reply by rarkins** [reply]
The downside is lack of visibility (in logs), but otherwise it doesn't make a functional change.

**Reply by ainame** [reply]
Yes, but looking at `lib/modules/manager/gradle/artifacts.ts` file, nothing uses `stdout` output in `Gradle` manager right now. Wouldn't my suggestion work fine?

https://github.com/renovatebot/renovate/blob/fa90477f6c54f0b8f3794afb54b16ce3f6e259a7/lib/modules/manager/gradle/artifacts.ts#L161

**Reply by rarkins** [reply]
Sure, go ahead

**Reply by ainame** [reply]
🚀 https://github.com/renovatebot/renovate/pull/20820
